File: ios/ios/Core/APIClient.swift
import Foundation

enum APIError: Error {
    case invalidResponse
    case unauthorized
    case unprocessable(String)
    case serverError
}

struct APIClient {
    private let authState: AuthState

    init(authState: AuthState) {
        self.authState = authState
    }

    func login(email: String, password: String) async throws -> String {
        var request = URLRequest(url: Config.apiBaseURL.appendingPathComponent("auth/token"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONEncoder().encode(["email": email, "password": password])

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }

        switch http.statusCode {
        case 200:
            let body = try JSONDecoder().decode([String: String].self, from: data)
            guard let token = body["token"] else { throw APIError.invalidResponse }
            return token
        case 422:
            let body = try? JSONDecoder().decode(ValidationErrorResponse.self, from: data)
            throw APIError.unprocessable(body?.firstMessage ?? "Invalid credentials.")
        default:
            throw APIError.serverError
        }
    }

    func register(name: String, email: String, password: String, passwordConfirmation: String) async throws {
        var request = URLRequest(url: Config.apiBaseURL.appendingPathComponent("auth/register"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONEncoder().encode([
            "name": name,
            "email": email,
            "password": password,
            "password_confirmation": passwordConfirmation,
        ])

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }

        switch http.statusCode {
        case 201:
            return
        case 422:
            let body = try? JSONDecoder().decode(ValidationErrorResponse.self, from: data)
            throw APIError.unprocessable(body?.firstMessage ?? "Registration failed.")
        default:
            throw APIError.serverError
        }
    }

    func fetchUser() async throws -> UserResponse {
        guard let token = authState.token else { throw APIError.unauthorized }

        var request = URLRequest(url: Config.apiBaseURL.appendingPathComponent("user"))
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        guard http.statusCode == 200 else { throw APIError.serverError }

        return try JSONDecoder().decode(UserResponse.self, from: data)
    }
}

struct UserResponse: Decodable {
    let id: Int
    let name: String
    let email: String
}

private struct ValidationErrorResponse: Decodable {
    let errors: [String: [String]]

    var firstMessage: String? {
        errors.values.first?.first
    }
}

File: android/app/src/main/java/com/tenuto/android/core/ApiClient.kt
package com.tenuto.android.core

import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST

@Serializable
private data class LoginRequest(val email: String, val password: String)

@Serializable
private data class RegisterRequest(
    val name: String,
    val email: String,
    val password: String,
    val password_confirmation: String,
)

@Serializable
private data class GoogleSignInRequest(val id_token: String)

@Serializable
data class TokenResponse(val token: String)

@Serializable
data class UserResponse(val id: Int, val name: String, val email: String)

@Serializable
private data class ValidationErrorResponse(val errors: Map<String, List<String>>) {
    val firstMessage: String? get() = errors.values.firstOrNull()?.firstOrNull()
}

sealed class ApiError : Exception() {
    data class Unprocessable(val apiMessage: String) : ApiError()
    data object InvalidResponse : ApiError()
    data object ServerError : ApiError()
    data object Unauthorized : ApiError()
}

private interface AuthService {
    @POST("auth/token")
    suspend fun login(@Body request: LoginRequest): Response<TokenResponse>

    @POST("auth/register")
    suspend fun register(@Body request: RegisterRequest): Response<Unit>

    @POST("auth/google/token")
    suspend fun googleSignIn(@Body request: GoogleSignInRequest): Response<TokenResponse>

    @GET("user")
    suspend fun fetchUser(@Header("Authorization") auth: String): Response<UserResponse>
}

@OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)
class ApiClient {
    private val json = Json { ignoreUnknownKeys = true }

    private val retrofit = Retrofit.Builder()
        .baseUrl(Config.apiBaseUrl)
        .client(OkHttpClient())
        .addConverterFactory(json.asConverterFactory("application/json; charset=UTF-8".toMediaType()))
        .build()

    private val service = retrofit.create(AuthService::class.java)

    suspend fun login(email: String, password: String): String {
        val response = service.login(LoginRequest(email, password))
        return when (response.code()) {
            200 -> response.body()?.token ?: throw ApiError.InvalidResponse
            422 -> throw ApiError.Unprocessable(parseValidationError(response.errorBody()?.string()) ?: "Invalid credentials.")
            else -> throw ApiError.ServerError
        }
    }

    suspend fun register(name: String, email: String, password: String, passwordConfirmation: String) {
        val response = service.register(RegisterRequest(name, email, password, passwordConfirmation))
        when (response.code()) {
            201 -> return
            422 -> throw ApiError.Unprocessable(parseValidationError(response.errorBody()?.string()) ?: "Registration failed.")
            else -> throw ApiError.ServerError
        }
    }

    suspend fun googleSignIn(idToken: String): String {
        val response = service.googleSignIn(GoogleSignInRequest(idToken))
        return when (response.code()) {
            200 -> response.body()?.token ?: throw ApiError.InvalidResponse
            422 -> throw ApiError.Unprocessable(parseValidationError(response.errorBody()?.string()) ?: "Google sign-in failed.")
            else -> throw ApiError.ServerError
        }
    }

    suspend fun fetchUser(token: String): UserResponse {
        val response = service.fetchUser("Bearer $token")
        return when (response.code()) {
            200 -> response.body() ?: throw ApiError.InvalidResponse
            401 -> throw ApiError.Unauthorized
            else -> throw ApiError.ServerError
        }
    }

    private fun parseValidationError(body: String?): String? {
        if (body == null) return null
        return try {
            json.decodeFromString<ValidationErrorResponse>(body).firstMessage
        } catch (e: Exception) {
            null
        }
    }
}
